mono/profiler/proflog.c

```diff
@@ -2490,6 +2541,517 @@ counters_and_perfcounters_sample (MonoProfiler *prof)
+	}
+
+	mono_mutex_lock (&coverage_mutex);
+
+	// Don't need to do anything else if we're already tracking this method
+	if (g_hash_table_lookup (coverage_methods, method)) {
+		COVERAGE_DEBUG(fprintf (stderr, "   Already tracking\n");)
+		mono_mutex_unlock (&coverage_mutex);
+		return TRUE;
+	}
+
+	klass = mono_method_get_class (method);
+	image = mono_class_get_image (klass);
+
+	// Don't handle coverage for the core assemblies
+	if (g_hash_table_lookup (suppressed_assemblies, mono_image_get_name (image)) != NULL) {
+		mono_mutex_unlock (&coverage_mutex);
+		return FALSE;
+	}
+
+	if (prof->coverage_filters) {
+		/* Check already filtered classes first */
+		if (g_hash_table_lookup (filtered_classes, klass)) {
+			COVERAGE_DEBUG(fprintf (stderr, "   Already filtered\n");)
+			mono_mutex_unlock (&coverage_mutex);
+			return FALSE;
+		}
+
+		classname = mono_type_get_name (mono_class_get_type (klass));
+
+		fqn = g_strdup_printf ("[%s]%s", mono_image_get_name (image), classname);
+
+		COVERAGE_DEBUG(fprintf (stderr, "   Looking for %s in filter\n", fqn);)
+		// Check positive filters first
+		has_positive = FALSE;
+		found = FALSE;
+		for (guint i = 0; i < prof->coverage_filters->len; ++i) {
+			char *filter = g_ptr_array_index (prof->coverage_filters, i);
+			if (filter [0] == '+') {
+				filter = &filter [1];
+				COVERAGE_DEBUG(fprintf (stderr, "   Checking against +%s ...", filter);)
+				if (strstr (fqn, filter) != NULL) {
+					COVERAGE_DEBUG(fprintf (stderr, "matched\n");)
+					found = TRUE;
+				} else {
+					COVERAGE_DEBUG(fprintf (stderr, "no match\n");)
+				}
+				has_positive = TRUE;
+			}
+		}
+		if (has_positive && !found) {
+			COVERAGE_DEBUG(fprintf (stderr, "   Positive match was not found\n");)
+			g_hash_table_insert (filtered_classes, klass, klass);
+			mono_mutex_unlock (&coverage_mutex);
+			g_free (fqn);
+			g_free (classname);
+			return FALSE;
+		}
+
+		for (guint i = 0; i < prof->coverage_filters->len; ++i) {
+			// Is substring search suffices ???
+			char *filter = g_ptr_array_index (prof->coverage_filters, i);
+			if (filter [0] == '+')
+				continue;
+
+			// Skip '-'
+			filter = &filter [1];
+			COVERAGE_DEBUG(fprintf (stderr, "   Checking against -%s ...", filter);)
+			if (strstr (fqn, filter) != NULL) {
+				COVERAGE_DEBUG(fprintf (stderr, "matched\n");)
+				g_hash_table_insert (filtered_classes, klass, klass);
+				mono_mutex_unlock (&coverage_mutex);
+				g_free (fqn);
+				g_free (classname);
+				return FALSE;
+			} else {
+				COVERAGE_DEBUG(fprintf (stderr, "no match\n");)
+			}
+		}
+		g_free (fqn);
+		g_free (classname);
+	}
+
+	COVERAGE_DEBUG(fprintf (stderr, "   Handling coverage for %s\n", mono_method_get_name (method));)
+	header = mono_method_get_header (method);
+
+	mono_method_header_get_code (header, &code_size, NULL);
+	if (code_size > 20000) {
+		g_warning ("Unable to instrument method %s:%s since it is too complex.", mono_class_get_name (klass), mono_method_get_name (method));
```

Comment:
Unless we actually fail to instrument it, I don't think we need to keep this warning.